osh/builtin_meta.py

```diff
@@ -75,66 +78,89 @@ def Run(self, cmd_val):
 
-        try:
-            f = self.fd_state.Open(resolved)  # Shell can't use descriptors 3-9
-        except (IOError, OSError) as e:
-            self.errfmt.Print_('source %r failed: %s' %
-                               (path, pyutil.strerror(e)),
-                               blame_loc=cmd_val.arg_locs[1])
-            return 1
+        else:
+            resolved = self.search_path.Lookup(path, exec_required=False)
+            if resolved is None:
+                resolved = path
 
-        line_reader = reader.FileLineReader(f, self.arena)
-        c_parser = self.parse_ctx.MakeOshParser(line_reader)
+            try:
+                f = self.fd_state.Open(resolved)  # Shell can't use descriptors 3-9
+            except (IOError, OSError) as e:
+                self.errfmt.Print_('source %r failed: %s' %
+                                   (path, pyutil.strerror(e)),
+                                   blame_loc=cmd_val.arg_locs[1])
+                return 1
+
+            line_reader = reader.FileLineReader(f, self.arena)
+            c_parser = self.parse_ctx.MakeOshParser(line_reader)
+
+            with process.ctx_FileCloser(f):
+                return self._Exec(cmd_val, arg_r, path, c_parser)
+
+    def _Exec(self, cmd_val, arg_r, path, c_parser):
+        # type: (cmd_value.Argv, args.Reader, str, CommandParser) -> int
+        call_loc = cmd_val.arg_locs[0]
 
         # A sourced module CAN have a new arguments array, but it always shares
         # the same variable scope as the caller.  The caller could be at either a
         # global or a local scope.
 
         # TODO: I wonder if we compose the enter/exit methods more easily.
-        with process.ctx_FileCloser(f):
-            with dev.ctx_Tracer(self.tracer, 'source', cmd_val.argv):
-                source_argv = arg_r.Rest()
-                with state.ctx_Source(self.mem, path, source_argv):
-                    with state.ctx_ThisDir(self.mem, path):
-                        src = source.SourcedFile(path, call_loc)
-                        with alloc.ctx_Location(self.arena, src):
-                            try:
-                                status = main_loop.Batch(
-                                    self.cmd_ev,
-                                    c_parser,
-                                    self.errfmt,
-                                    cmd_flags=cmd_eval.RaiseControlFlow)
-                            except vm.IntControlFlow as e:
-                                if e.IsReturn():
-                                    status = e.StatusCode()
-                                else:
-                                    raise
+
+        with dev.ctx_Tracer(self.tracer, 'source', cmd_val.argv):
+            source_argv = arg_r.Rest()
+            with state.ctx_Source(self.mem, path, source_argv):
+                with state.ctx_ThisDir(self.mem, path):
+                    src = source.SourcedFile(path, call_loc)
```

Comment:
I think we should change this to 2 options

- `source.DiskFile()`
- `source.BuiltinFile()` 

or something.  And make the description in errors nicer.   It will be confusing for users if they try to find "math.ysh".


But I guess we can do that when we make a pass over error messages